src/ray/common/cgroup/cgroup_utils.cc

```diff
@@ -0,0 +1,163 @@
+#include "absl/strings/str_split.h"
+#include "ray/util/logging.h"
+
+namespace ray {
+
+namespace {
+
+// Owner can read and write.
+constexpr int kCgroupFilePerm = 0600;
+
+// Open a cgroup path and append write [content] into the file.
+void OpenCgroupFileAndAppend(std::string_view path, std::string content) {
+  std::ofstream out_file{path.data(), std::ios::out | std::ios::app};
+  out_file << content;
+}
+
+bool CreateNewCgroup(const PhysicalModeExecutionContext &ctx) {
+  // Sanity check.
+  RAY_CHECK(!ctx.uuid.empty());
+  RAY_CHECK_NE(ctx.uuid, kDefaultCgroupUuid);
+  RAY_CHECK_GT(ctx.max_memory, 0);
+
+  const std::string cgroup_folder =
+      absl::StrFormat("%s/%s", ctx.cgroup_directory, ctx.uuid);
+  int ret_code = mkdir(cgroup_folder.data(), kCgroupFilePerm);
+  if (ret_code != 0) {
+    return false;
+  }
+
+  if (ctx.max_memory > 0) {
+    const std::string procs_path = absl::StrFormat("%s/cgroup.procs", cgroup_folder);
+    OpenCgroupFileAndAppend(procs_path, absl::StrFormat("%d", ctx.pid));
+
+    const std::string max_memory_path = absl::StrFormat("%s/memory.max", cgroup_folder);
+    OpenCgroupFileAndAppend(max_memory_path, absl::StrFormat("%d", ctx.max_memory));
+  }
+
+  return true;
+}
+
+bool UpdateDefaultCgroup(const PhysicalModeExecutionContext &ctx) {
+  // Sanity check.
+  RAY_CHECK(!ctx.uuid.empty());
+  RAY_CHECK_EQ(ctx.uuid, kDefaultCgroupUuid);
+  RAY_CHECK_EQ(ctx.max_memory, 0);
+
+  const std::string cgroup_folder =
+      absl::StrFormat("%s/%s", ctx.cgroup_directory, ctx.uuid);
+  int ret_code = mkdir(cgroup_folder.data(), kCgroupFilePerm);
+  if (ret_code != 0) {
+    return false;
+  }
+
+  const std::string procs_path = absl::StrFormat("%s/cgroup.procs", cgroup_folder);
+  OpenCgroupFileAndAppend(procs_path, absl::StrFormat("%d", ctx.pid));
+
+  return true;
+}
+
+bool DeleteCgroup(const PhysicalModeExecutionContext &ctx) {
+  // Sanity check.
+  RAY_CHECK(!ctx.uuid.empty());
+  RAY_CHECK_NE(ctx.uuid, kDefaultCgroupUuid);
+  RAY_CHECK_GT(ctx.max_memory, 0);
+
+  const std::string cgroup_folder =
+      absl::StrFormat("%s/%s", ctx.cgroup_directory, ctx.uuid);
+  return rmdir(cgroup_folder.data()) == 0;
+}
+
+bool RemoveCtxFromDefaultCgroup(const PhysicalModeExecutionContext &ctx) {
+  // Sanity check.
+  RAY_CHECK(!ctx.uuid.empty());
+  RAY_CHECK_EQ(ctx.uuid, kDefaultCgroupUuid);
+  RAY_CHECK_EQ(ctx.max_memory, 0);
+
+  const std::string cgroup_folder =
+      absl::StrFormat("%s/%s", ctx.cgroup_directory, ctx.uuid);
+  int ret_code = mkdir(cgroup_folder.data(), kCgroupFilePerm);
+  if (ret_code != 0) {
+    return false;
+  }
+
+  const std::string procs_path = absl::StrFormat("%s/cgroup.procs", cgroup_folder);
+  std::ostringstream buffer;
+  {
+    std::ifstream file(procs_path.data(), std::ios::in);
+    buffer << file.rdbuf();
+  }
+  std::string content = buffer.str();  // contains all PIDs, separated by space
+
+  std::vector<std::string_view> old_pid_strings = absl::StrSplit(content, ' ');
```

Comment:
You don't need to remove the proc from the previous cgroup. If you put it into a new cgroup, it's auto removed from the old cgroup.

Reply:
> If you put it into a new cgroup, it's auto removed from the old cgroup.

I rewrite the whole cgroup file content; the cgroup is of a fixed cgroup path.

Reply:
I think you are still reading N pids, and write back N-1 pids?